core/common/src/main/java/alluxio/metrics/MetricsSystem.java

```diff
@@ -341,12 +342,17 @@ public static String getJobWorkerMetricName(String name) {
    * @return the metric registry name
    */
   private static String getMetricNameWithUniqueId(InstanceType instance, String name) {
+    AlluxioConfiguration conf = new InstancedConfiguration(ConfigurationUtils.defaults());
+    String identifier;
+    if (conf.isSet(PropertyKey.USER_APP_ID)) {
+      identifier = conf.get(PropertyKey.USER_APP_ID);
+    } else {
+      identifier = NetworkAddressUtils.getLocalHostMetricName(sResolveTimeout);
```

Comment:
In the getLocalHostMetricName, we have the caching logic, maybe better to put this logic with caching.

In addition, this function is also used by instanceType=WORKER but the changes may relate to clients only.

I still no sure is it right to change it.

The `metrics.Metric.java` class may also need to be changed so that metric names can be processed correctly with the new format. Metric name is so tricky.....!